feat: 14414 Replace redux-persist with barebones JS persist system using FileSystemStorage

MarioAslau opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

This PR aims at replacing our redux-persist with barebones JS Persistence using FileSystemStorage.

As a first step, all controller state is being removed from the regular redux-persist into FileSystemStorage, with each controller being written to a separate file vs storing the whole Engine object into one file.

**EngineService**:
* setupEnginePersistence:
  * Acts as the bridge between the Engine's controller system and the storage layer, setting up real-time persistence for individual controllers
  * Contains Redux synchronization by calling this.updateBatcher.add(controllerName) to efficiently batch Redux updates after successful persistence

**persistConfig** now has:
* ControllerStorage methods (getter and setter) for persisted controllers in FilesystemStorage
* Note: The existing persistUserTransform and persistOnboardingTransform remain for non-engine state; engine state is now blacklisted from redux-persist

**Migrations** now have:
* To keep developer experience the same when creating migrations, the migration system now includes:
  * deflateToControllersAndStrip → "repacking" the single object back into distributed files and safely preserving failed controllers to prevent data loss
  * inflateFromControllers → "unpacking" distributed files back into a single object for migration processing
* Migration 104: Checks if old persisted engine state is present, migrates each controller to individual storage files, then removes successfully migrated controllers from the old redux-persist state (preserving any that failed to migrate)

**Key Benefits:**
* ndividual controller failures don't affect other controllers
* Performance: Smaller, focused file operations instead of large monolithic writes
* Data Safety: Critical bug fix prevents silent data loss during migration failures

Note: if .env file value is not set right, this will trigger a Vault Recovery on dev environments after locking and reopening of the application. Due to this conditions we have for E2E: const state = isE2E ? reduxState?.engine?.backgroundState  : persistedState?.backgroundState ?? {};


## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: 

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.





<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Replaces engine redux-persist with per-controller filesystem storage, wires EngineService to persist and batch controller updates, and adds migration 104 with inflate/deflate helpers plus tests and docs.
> 
> - **State Persistence (core change)**
>   - Introduces `ControllerStorage` and `createPersistController` in `store/persistConfig/index.ts` for per-controller FilesystemStorage persistence (debounced writes).
>   - Blacklists `engine` from redux-persist; keeps `user` and `onboarding` transforms only.
>   - Adds `ControllerStorage.getAllPersistedState()` to load controller states at startup.
> - **EngineService**
>   - `start()` now prefers persisted controller state (non-E2E) and initializes controllers via `initializeControllers`.
>   - `setupEnginePersistence()` subscribes to `BACKGROUND_STATE_CHANGE_EVENT_NAMES`, batches Redux updates (`UPDATE_BG_STATE_KEY`), and persists filtered controller state.
>   - Renames `updateControllers` → `initializeControllers`; enhances batching logic.
> - **Migrations**
>   - Adds `104` to split legacy `engine.backgroundState` into per-controller files and preserve failed controllers.
>   - Updates `migrations/index.ts` to inflate controller files before 105+ and deflate back after last migration; strips `engine` only if all saves succeed.
> - **Tests & Tooling**
>   - Replaces old `persistConfig` tests; adds comprehensive tests for new persistence, migration 104, and EngineService (timers, batching, persistence, error handling).
>   - Updates test setup to mock `unstable_batchedUpdates` and related dependencies.
> - **Docs**
>   - Adds `STATE_FLOW_DOCUMENTATION.md` describing end-to-end state flow and new persistence architecture.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 60d270a32a6dee0ded1ac40069e2c0818eaa2540. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

MarioAslau committed:
feat: Fix to persist Biometrics on manual lock

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: partial fix

MarioAslau committed:
Merge branch 'feat/14414-persist-biometrics' of https://github.com/MetaMask/metamask-mobile into feat/14414-persist-biometrics

MarioAslau committed:
feat: fix Biometrics Pop up bug

MarioAslau committed:
feat: fix prettier warning

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: smoke test run_tag_smoke_identity_android

MarioAslau committed:
Merge branch 'feat/14414-persist-biometrics' of https://github.com/MetaMask/metamask-mobile into feat/14414-persist-biometrics

MarioAslau committed:
feat: e2e tweaks for failing test

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: fix lint warning

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: fix unit test

MarioAslau committed:
feat: formatting

MarioAslau committed:
feat: change string to const value

MarioAslau committed:
feat: e2e tweaks

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: revert viewHelper

MarioAslau committed:
feat: Increase App Unit tests coverage

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: update unit tests and fix lint errors

MarioAslau committed:
feat: Increase Sonar Cloud Coverage

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
fix: linter and unit test

MarioAslau committed:
feat: Fix CI lint issue

MarioAslau committed:
feat: writing Controller state in new persist & initialising Engine with it

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: include Engine Tranform filterin in new Persistance mechanism

MarioAslau committed:
feat: Consolidate Controller subscription + migration 91

MarioAslau committed:
feat: migration test

MarioAslau committed:
feat: Fixing Unti tests for persistConfig/migrations

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: fixing unit tests for EngineService + adressing Cursor bot comments

MarioAslau commented on app/store/migrations/091.ts:
```diff
@@ -0,0 +1,129 @@
+            const key = `persist:${controllerName}`;
+            const value = JSON.stringify(controllerState);
+
+            await FilesystemStorage.setItem(key, value, Device.isIos());
```
This is wrong. The pre-existing data in the old storage is already filtered when it was stored there. Adding an additional filter inside the migration would be re-filtering filtered data. It is redundant.

MarioAslau committed:
feat: fix OptinMetrics test

MarioAslau committed:
feat: reverse snapshot hange

MarioAslau committed:
feat: fixing test

MarioAslau committed:
feat: update unit test

MarioAslau committed:
feat: fixing Unit tests testSetup

MarioAslau committed:
feat: update unit test...again

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: Update snapshot

MarioAslau committed:
feat: Mock unstable_batchedUpdates directly in the react-native module

MarioAslau committed:
feat: updating tests

MarioAslau committed:
feat: Adressing Cursor Bot comment

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: fox linting issues and migration test import

MarioAslau committed:
feat:  fix persistConfig test

MarioAslau committed:
feat: fix Cursor Bot migration 092 uncovered case

Cal-L commented on app/store/persistConfig.ts:
```diff
@@ -13,6 +13,122 @@ import { getPersistentState } from './getPersistentState/getPersistentState';
+              const { _persist, ...controllerState } = parsedData;
+
+              // Only include controllers that have meaningful state (not empty objects)
+              if (Object.keys(controllerState).length > 0) {
```
Let's verify if it's possible to have a controller that exists in persistence without state

Cal-L replied:
We can probably remove this condition to save on performance since this isn't likely to break things

Cal-L commented on app/store/persistConfig.ts:
```diff
@@ -13,6 +13,122 @@ import { getPersistentState } from './getPersistentState/getPersistentState';
+      );
+
+      // Combine all controller states into a single object, filtering out null values
+      const backgroundState = controllerStates
```
Instead of post processing controllerStates, you can just construct the object as it goes through the mapping. Then you can just return controllerStates here instead

Cal-L commented on app/store/persistConfig.ts:
```diff
@@ -59,57 +175,56 @@ const MigratedStorage = {
-        value,
-        // @ts-expect-error - EngineContext have stateless controllers, so metadata is not available
-        Engine.context[key as keyof EngineContext]?.metadata,
+              ReduxService.store.dispatch({
```
Let's add some comments to indicate what this is doing

Cal-L commented on app/store/persistConfig.ts:
```diff
@@ -59,57 +175,56 @@ const MigratedStorage = {
-      // Engine not initialized, skipping transform
-      return inboundState;
-    }
+              await ControllerStorage.setItem(
```
Let's add some comments to indicate what this is doing

Cal-L reviewed (commented):
Left some comments

MarioAslau committed:
feat: addressing PR comments

MarioAslau committed:
feat: add comments to setupEnginePersistence for clarity

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau commented on app/store/persistConfig/index.ts:
```diff
@@ -0,0 +1,245 @@
+                `Failed to persist ${controllerName} state`,
+              );
+            }
+          }, 200),
```
The race condition is less critical than underlined above:
- Individual controller scope: Each controller has its own debounced function
- Sequential execution within controller: Each execution waits for completion
- File overwrites are atomic: Filesystem writes are typically atomic operations
- Dispatch happens after persistence
- Only issue: rapid state changes might lose intermediate updates (but that's expected with debouncing)

The race condition exists but might be more of a theoretical issue than a practical problem in this specific case.

MarioAslau committed:
feat: increase test coverage in persistConfig for SoundCloud

metamaskbot committed:
Bump version number to 2257

metamaskbot committed:
Bump version number to 2258

MarioAslau committed:
feat: fix for failing Fixtures

MarioAslau committed:
Merge branch 'feat/14414-persist-biometrics' of https://github.com/MetaMask/metamask-mobile into feat/14414-persist-biometrics

MarioAslau committed:
feat: update state conditional

MarioAslau committed:
feat: bump migration version

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau commented:
Bitrise tests before merging Main: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/7a46bb54-209b-4d15-9151-8983f2d66b07

MarioAslau committed:
feat: push migration to 96

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: push migration number

MarioAslau committed:
feat: merge main

MarioAslau committed:
COMMIT

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: remove isTest & update Controller Constants

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: 2276 New Migration Pattern to support Persistance Splitting (#18965)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

Due to our Engine State being split from the old redux-persist system in
this https://github.com/MetaMask/metamask-mobile/pull/17685 , it was
moved into a new controler per file based system.

We need to establish a new way to take into account future migrations
that might deal with state changes on the Engine Controllers States as
we now store our persisted state in two systems, the original one using
redux-persist and the barebones one, that is currently storing all the
Controllers.

In order to keep Developer experience the same, moving forward, any new
migrations will be added and done in the same way. All post migrations
<= 98 will be using the same old system, migrations post-98 will be
using the new logic. State is pulled from both persistance systems,
inflating the state, made into the old shape. Any new migrations will
change this object. After that, the state is deflated and split into
it's respective persistance systems.

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes:

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->



https://github.com/user-attachments/assets/d83750ef-fd9f-4b0c-a9c5-33a7fcdc0334



## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

MarioAslau committed:
feat: bump migration numbers

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
fix: correct migration 101 test imports and references

MarioAslau committed:
feat: fix migration numbers

metamaskbot committed:
[skip ci] Bump version number to 2528

MarioAslau committed:
feat: fix unit test for useAccountInfo

MarioAslau committed:
Merge branch 'feat/14414-persist-biometrics' of https://github.com/MetaMask/metamask-mobile into feat/14414-persist-biometrics

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: fix useAccountInfo

MarioAslau committed:
feat: increase minimumVersion for useAccountInfo

MarioAslau committed:
feat: bump migrations again

MarioAslau committed:
feat: bump migration test number

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

metamaskbot committed:
[skip ci] Bump version number to 2536

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: removal of blank line

MarioAslau committed:
Merge branch 'feat/14414-persist-biometrics' of https://github.com/MetaMask/metamask-mobile into feat/14414-persist-biometrics

metamaskbot committed:
[skip ci] Bump version number to 2554

tommasini committed:
merge main and solve conflicts

metamaskbot committed:
[skip ci] Bump version number to 2563

MarioAslau committed:
feat: fix migration 103 test file import

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: update unused import

MarioAslau committed:
feat: fix migration import name

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: fix for migration103 test

MarioAslau committed:
feat: possible chainId hex fix

MarioAslau committed:
feat: bump migrations again

MarioAslau committed:
feat: bump migration number for deflation

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: more hex safe guarding

MarioAslau committed:
feat: remove hex safeguarding

MarioAslau commented on app/core/EngineService/EngineService.ts:
```diff
@@ -72,7 +72,11 @@ export class EngineService {
     });
-    const state = reduxState?.engine?.backgroundState ?? {};
+
+    const state = isE2E
```
this isE2E can cause problems when developing locally if the env variable is not set/read well, which will result in vault recovery after re-opening the app.


Comment it out if causing issues and use  'const state = persistedState?.backgroundState ?? {};'

tommasini committed:
merge main and solve conflicts

andrepimenta committed:
Merge branch 'main' into feat/14414-persist-biometrics

andrepimenta committed:
Fix/controller storage refactor (#20635)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

Fix controller redux change.

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes:

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

tommasini committed:
Merge branch 'main' into feat/14414-persist-biometrics

tommasini committed:
cleanup code

tommasini committed:
fix unit tests

tommasini committed:
Merge branch 'main' into feat/14414-persist-biometrics

tommasini committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: Fix migration number for inflate/deflate

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

metamaskbot committed:
[skip ci] Bump version number to 2667

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau commented on app/store/persistConfig/index.ts:
```diff
@@ -0,0 +1,268 @@
+                payload: { key: controllerName },
+              });
+
+              await persistController(filteredState);
```
This is false, redux is primary and should be executed immediately, persisting comes secondary and can be done after.

MarioAslau commented on app/store/persistConfig/index.ts:
```diff
@@ -0,0 +1,268 @@
+                `Failed to process ${controllerName} state change`,
+              );
+            }
+          },
```
Same reasoning as the others. Redux State is primary and the persist is secondary, coming after.

MarioAslau commented on app/store/persistConfig/index.ts:
```diff
@@ -0,0 +1,268 @@
+                payload: { key: controllerName },
+              });
+
+              await persistController(filteredState);
```
Redux should have priority, persitence coming secondary after

MarioAslau commented on app/store/migrations/104.test.ts:
```diff
@@ -0,0 +1,213 @@
+jest.mock('../../util/device');
+const mockDevice = Device as jest.Mocked<typeof Device>;
+
+describe('Migration 103', () => {
```
stale, file has been updated

MarioAslau commented on app/store/migrations/index.ts:
```diff
@@ -224,37 +226,124 @@ export const migrationList: MigrationsList = {
           onMigrationsComplete(migratedState);
         }
 
+        if (Number(migrationNumber) === lastVersion && lastVersion >= 102) {
```
outdated

MarioAslau committed:
feat: clean up code

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau commented:
Bitrise PR E2E: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/a6377d95-9726-48d9-a648-b2cd1c312977

metamaskbot committed:
[skip ci] Bump version number to 2675

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

Cal-L reviewed (commented):
Let's resolve the cursor comments and add JS docs in the changes for the migrations/index file

metamaskbot committed:
[skip ci] Bump version number to 2689

MarioAslau committed:
feat: addressing PR changes

MarioAslau committed:
feat: format persistConfig

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: bump version

metamaskbot committed:
[skip ci] Bump version number to 2693

metamaskbot committed:
[skip ci] Bump version number to 2694

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: use batcher for updates & consolidate Storages

MarioAslau committed:
feat: Unit test fixes & lint fix

metamaskbot committed:
[skip ci] Bump version number to 2695

MarioAslau committed:
feat: addressing lint check

MarioAslau committed:
Merge branch 'feat/14414-persist-biometrics' of https://github.com/MetaMask/metamask-mobile into feat/14414-persist-biometrics

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: lint/ts check fix in EngineService

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: PR updates

metamaskbot committed:
[skip ci] Bump version number to 2699

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

joaoloureirop committed:
fix lint issues

joaoloureirop committed:
Merge branch 'main' into feat/14414-persist-biometrics

joaoloureirop committed:
remove formatter changes

MarioAslau committed:
feat: lint error fix - removed unused import

MarioAslau committed:
feat: state management flow documentation

MarioAslau committed:
feat: improved failed controller tracking and avoid data loss in case of failed migration

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

MarioAslau committed:
feat: revert gradle version

MarioAslau committed:
Merge branch 'feat/14414-persist-biometrics' of https://github.com/MetaMask/metamask-mobile into feat/14414-persist-biometrics

MarioAslau committed:
feat: fix lint issue in migration by removing else

metamaskbot committed:
[skip ci] Bump version number to 2700

MarioAslau committed:
Merge branch 'main' into feat/14414-persist-biometrics

joaoloureirop committed:
state fallback for e2e mode as well

MarioAslau committed:
feat: revert grade build

joaoloureirop committed:
remove version bumps

asalsys reviewed (commented):
I would update the ControllerStorage naming to StorageController
